pika/adapters/select_connection.py

```diff
@@ -188,23 +135,54 @@ class SelectPoller(object):
     to override the update_handler and start methods for additional types.
 
     """
-    TIMEOUT = 0.01
+    # Drop out of the poll loop every POLL_TIMEOUT secs as a worst case, this
+    # is only a backstop value. We will run timeouts when they are scheduled.
+    POLL_TIMEOUT = 5
+    # if the poller uses MS specify 1000
+    POLL_TIMEOUT_MULT = 1
 
-    def __init__(self, fileno, handler, events, state_manager):
+    def __init__(self):
         """Create an instance of the SelectPoller
 
-        :param int fileno: The file descriptor to check events for
-        :param method handler: What is called when an event happens
-        :param int events: The events to look for
-        :param method state_manager: The method to manage state
-
         """
-        self.fileno = fileno
-        self.events = events
-        self.open = True
-        self._handler = handler
-        self._timeouts = []
-        self._manage_event_state = state_manager
+        self._fd_handlers = dict()
+        self._fd_events = {READ: set(), WRITE: set(), ERROR: set()}
+        self._stopping = False
+        self._timeouts = {}
+        self._next_timeout = None
+        self._processing_fd_event_map = {}
+        self._r_interrupt, self._w_interrupt = self.get_interrupt_pair()
+        self.add_handler(self._r_interrupt.fileno(), self.read_interrupt, READ)
+
+    def get_interrupt_pair(self):
+        """ Use a socketpair to be able to interrupt the ioloop if called
+            from another thread. Socketpair() is not supported on some OS (Win)
+            so use a pair of simple UDP sockets instead. The sockets will be
+            closed and garbage collected by python when the ioloop itself is.
+        """
+        try:
+            return socket.socketpair()
+
+        except:
+            LOGGER.debug("Using custom socketpair for interrupt")
+            read_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
```

Comment:
No, it doesn't. But then the socketpair is basically an optimisation to
drop out of the ioloop immediately when triggered from elsewhere. It will
work for 99.99% of cases and, as long as the default timeout is kept at
around 5 secs, it's unlikely to be an issue there. I didn't want to use a
tcp socket because then we have to handle the connect/accept cycle and the
potential to block or fail there. I'd rather we just lose the optimisation.
On 4 May 2015 22:34, "vitaly-krugl" <notifications@github.com> wrote:

> In pika/adapters/select_connection.py
> <https://github.com/pika/pika/pull/555#discussion_r29626106>:
>
> > +        self._processing_fd_event_map = {}
> > +        self._r_interrupt, self._w_interrupt = self.get_interrupt_pair()
> > +        self.add_handler(self._r_interrupt.fileno(), self.read_interrupt, READ)
> > +
> > +    def get_interrupt_pair(self):
> > +        """ Use a socketpair to be able to interrupt the ioloop if called
> > +            from another thread. Socketpair() is not supported on some OS (Win)
> > +            so use a pair of simple UDP sockets instead. The sockets will be
> > +            closed and garbage collected by python when the ioloop itself is.
> > +        """
> > +        try:
> > +            return socket.socketpair()
> > +
> > +        except:
> > +            LOGGER.debug("Using custom socketpair for interrupt")
> > +            read_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
>
> @wjps <https://github.com/wjps>, this may be a bug. My understanding is
> that UDP does not guarantee packet delivery, possibly not even on the
> loopback interface.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pika/pika/pull/555/files#r29626106>.
>